analyses/collapse-rnaseq/cwl/scripts/liftover_collapse_rnaseq.py

```diff
@@ -0,0 +1,177 @@
+    )
+    parser.add_argument(
+        "-g", "--gene-id", action="store", dest="gene_id", help="NAME of gene ID (ENSG values) column"
+    )
+    parser.add_argument(
+        "-n", "--gene-name", action="store", dest="gene_name", help="NAME of gene name column, if present to keep if not found"
+    )
+    parser.add_argument(
+        "-s", "--skip", action="store", dest="skip", type=int, help="Number of lines to skip if needed"
+    )
+    parser.add_argument(
+        "-c", "--collapse", action="store_true", dest="collapse", help="If set, will collapse on gene_symbol"
+    )
+
+    args = parser.parse_args()
+
+    gtf_dict = {}
+    # hardcoded for now, could make opts
+    tx_type = 2
+    tx_desc = 8
+    # set up pattern matcher ahead of time for efficiency
+    id_find = re.compile(r"gene_id \"(ENSG\d+)\.\d+\";")
+    name_find = re.compile(r".*gene_name \"(\S+)\";")
+    print('Indexing gtf', file=sys.stderr)
+    with (gzip.open if args.gtf.endswith("gz") else open)(args.gtf, "rt", encoding="utf-8") as gtf:
+        for line in gtf:
+            if not line.startswith('##'):
+                update_gtf_dict(gtf_dict, line, tx_type, tx_desc, id_find, name_find)
+    # track sym occurrences if collapse flag given for later processing
+    gene_sym_ct_dict = {}
+    with (gzip.open if args.table.endswith("gz") else open)(args.table, "rt", encoding="utf-8") as table:
+        if args.skip:
+            print('Skipping ' + str(args.skip) + ' lines', file=sys.stderr)
+            for i in range(args.skip):
+                skip = next(table)
+        head = next(table)
+        header = head.rstrip('\n').split('\t')
+        # get positions of current gene id and name
+        g_idx = header.index(args.gene_id)
+        n_idx = header.index(args.gene_name)
+        # make assumption that data will start after ID and name
+        d_idx = max(g_idx, n_idx) + 1
+        print('Processing table', file=sys.stderr)
+        out_lift_fn = args.out + '.liftover.tsv'
+        out_lift = open(out_lift_fn, 'w')
+        # Some counters to track progress
+        n = 10000
+        l = 0
+        out_head = "{}\t{}\t{}\n".format('gene_id', 'gene_name', '\t'.join(header[d_idx:]))
+        out_lift.write(out_head)
+        for line in table:
+            if l % n == 0:
+                print("Processed " + str(l) + " lines", file=sys.stderr)
+            out_data = liftover(line, gene_sym_ct_dict, g_idx, n_idx, gtf_dict, d_idx)
+            out_lift.write(out_data)
+            l += 1
+        # dump out rest of leftover lines
+        out_lift.close()
+    
+    if args.collapse:
+        print("Collapse flag given. Collating dups from liftover", file=sys.stderr)
+        dup_gene_sym_dict = {}
+        dup_ct = 0
+        out_collapse_fn = args.out + '.collapsed.tsv'
+        out_collapse = open(out_collapse_fn, 'w')
+        with (gzip.open if out_lift_fn.endswith("gz") else open)(out_lift_fn, "rt", encoding="utf-8") as lifted:
```

Comment:
I guess its safe but we know exactly what this file is since we just made it moments ago. It's a unzipped TSV. The gzip check here is not necessary.